cadence/src/builder.rs

```diff
@@ -195,11 +236,49 @@ impl<'a> MetricFormatter<'a> {
         Self::TAG_PREFIX.len() + self.kv_size + self.tags.len() - 1
     }
 
+    fn timestamp_size_hint(&self) -> usize {
+        if let Some(_timestamp) = self.timestamp {
+            /* |T */
+            2 + /* timestamp */ 10
+        } else {
+            0
+        }
+    }
+
+    fn sampling_rate_size_hint(&self) -> usize {
+        if let Some(_rate) = self.sampling_rate {
+            /* |@ */
+            2 + /* rate */ 10
+        } else {
+            0
+        }
+    }
+
+    fn container_id_size_hint(&self) -> usize {
+        if let Some(container_id) = self.container_id {
+            /* |c */
+            2 + container_id.len()
+        } else {
+            0
+        }
+    }
+
+    fn size_hint(&self) -> usize {
+        self.base_size
+            + self.sampling_rate_size_hint()
+            + self.tag_size_hint()
+            + self.timestamp_size_hint()
+            + self.container_id_size_hint()
+    }
+
     pub(crate) fn format(&self) -> String {
-        let size_hint = self.base_size + self.tag_size_hint();
+        let size_hint = self.size_hint();
         let mut metric_string = String::with_capacity(size_hint);
         self.write_base_metric(&mut metric_string);
```

Comment:
Does the order of these matter? If so, could you add a test that uses all of the new fields and tags?

Reply:
I think the order doesn't matter https://github.com/DataDog/datadog-agent/blob/3ca96b63fbc790ec56c2a69223faf24d5aeb3fcb/comp/dogstatsd/server/parse.go#L184